packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -1037,31 +1158,58 @@ export async function runSpec(
   config: SpecConfig,
   steps: SpecStep[]
 ): Promise<void> {
-  let runner: TestRunner;
-  if (usePersistence) {
-    runner = new IndexedDbTestRunner(name, config);
-  } else {
-    runner = new MemoryTestRunner(name, config);
+  console.log('Running spec: ' + name);
+  const platform = new TestPlatform(PlatformSupport.getPlatform());
+  let runners: TestRunner[] = [];
+  for (let i = 0; i < config.numClients; ++i) {
+    if (usePersistence) {
+      runners.push(new IndexedDbTestRunner(name, platform, config));
+    } else {
+      runners.push(new MemoryTestRunner(name, platform, config));
+    }
+    await runners[i].start();
   }
-  await runner.start();
+  let lastStep = null;
+  let count = 0;
   try {
-    await runner.run(steps);
+    await sequence(steps, async step => {
+      ++count;
+      lastStep = step;
+      return runners[step.clientIndex || 0].run(step);
+    });
+  } catch (err) {
+    console.warn(
+      `Spec test failed at step ${count}: ${JSON.stringify(lastStep)}`
+    );
+    throw err;
   } finally {
-    await runner.shutdown();
+    for (const runner of runners) {
+      await runner.shutdown();
+    }
+    if (usePersistence) {
+      await IndexedDbTestRunner.destroyPersistence();
+    } else {
+      await MemoryTestRunner.destroyPersistence();
```

Comment:
Given we're no longer relying on TestRunner.destroyPersistence() as a virtual method implemented by both runners, you can probably just delete MemoryTestRunner.destroyPersistence().

Reply:
Removed.